auction/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use rstd::result;
+use sr_primitives::traits::{MaybeSerializeDeserialize, Member, SimpleArithmetic};
+use srml_support::{
+	decl_error, decl_event, decl_module, decl_storage, dispatch::Result, ensure, Parameter, StorageMap, StorageValue,
+};
+use srml_system::{self as system, ensure_signed};
+use utilities::{LinkedItem, LinkedList};
+
+use traits::{Auction, AuctionHandler, AuctionInfo, OnNewBidResult};
+
+mod mock;
+mod tests;
+
+pub trait Trait: srml_system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as srml_system::Trait>::Event>;
+	type Balance: Parameter + Member + SimpleArithmetic + Default + Copy + MaybeSerializeDeserialize;
+	type AuctionId: Parameter + Member + SimpleArithmetic + Default + Copy + MaybeSerializeDeserialize;
+	type Handler: AuctionHandler<Self::AccountId, Self::Balance, Self::BlockNumber, Self::AuctionId>;
+}
+
+type AuctionEndTimeList<T> =
+	LinkedList<AuctionEndTime<T>, <T as srml_system::Trait>::BlockNumber, <T as Trait>::AuctionId>;
+
+decl_event!(
+	pub enum Event<T> where
+		<T as srml_system::Trait>::AccountId,
+		<T as Trait>::Balance,
+		<T as Trait>::AuctionId,
+	{
+		Bid(AuctionId, AccountId, Balance),
+	}
+);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Auction {
+		pub Auctions get(fn auctions): map T::AuctionId => Option<AuctionInfo<T::AccountId, T::Balance, T::BlockNumber>>;
+		pub AuctionsCount get(fn auctions_count): T::AuctionId;
+		pub AuctionEndTime get(fn auction_end_time): map(T::BlockNumber, Option<T::AuctionId>) => Option<LinkedItem<T::AuctionId>>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		pub fn bid(origin, id: T::AuctionId, value: T::Balance) -> Result {
+			let from = ensure_signed(origin)?;
+
+			if let Some(mut auction) = <Auctions<T>>::get(id) {
+				if let Some(current_bid) = auction.bid.clone() {
```

Comment:
```suggestion
				if let Some(ref current_bid) = auction.bid {
```

or `if let Some(current_bid) = auction.bid.as_ref() {`